batch.go

```diff
@@ -92,6 +96,56 @@ func (ebi *EventBatchIterator) Next(eventID string, opts *BatchOptions) (*EventB
 	return eb, nil
 }
 
+// NextWithPartition fetches the batch of messages in the message slice at a position one larger than the last one accessed and then map them with partitionKey.
+func (ebi *EventBatchIterator) NextWithPartition(opts *BatchOptions) (map[string]*EventBatch, error) {
+	if ebi.Done() {
+		return nil, ErrNoMessages{}
+	}
+
+	if opts == nil {
+		opts = &BatchOptions{
+			MaxSize: DefaultMaxMessageSizeInBytes,
+		}
+	}
+
+	ebMap := make(map[string]*EventBatch)
```

Comment:
Perhaps, the building of the map should be done in the constructor of the batch iterator when the slice of events is provided by the consumer. I believe this would make the logic in `Next` simpler.

Reply:
Make sense. I am also not sure how to deal with events without `PartitionKey`. Is it necessary for users to set `PartitionKey` for each events in which case I need to throw exceptions? Or I just send these events to a random partition?

Reply:
Nil partitioned events should be in a batch themselves and be distributed round robin (default for not setting the partition key).

Probably want to have a nil key represented in the partition map of events.